mc-sema/runtime/linux_i386_callback.c

```diff
@@ -0,0 +1,736 @@
+#define COUNT_LEVEL(N) \
+  ".align 8, 0x90\n" \
+  #N ":" \
+  "incl %1\n"
+
+void do_call_value(void *state, uint32_t value)
+{
+    // get a clean frame to store state
+    int32_t prev_call_frame = cur_do_call_frame++;
+    do_call_state_t *cs = &(do_call_state[cur_do_call_frame]);
+    call_frame_counter = 0;
+    cs->__mcsema_jmp_count = NUM_DO_CALL_FRAMES - cur_do_call_frame - 1;
+
+    __asm__ volatile(
+            "pusha\n" // save all regs just so we don't have bother keeping track of what we saved
+            "fxsave %0\n" // save sse state
+            "movl %3, %%eax\n"  // capture "state" arg (mcsema regstate)
+            "movl %4, %%ecx\n"  // capture "value" arg (call destination)
+            "movl %2, %%esi\n" // pointer to TLS area where we save state
+            "movl %c[state_edi](%%eax), %%edi\n" // dump struct regs to state
+            "movl %c[state_edx](%%eax), %%edx\n"
+            "movl %c[state_ebx](%%eax), %%ebx\n"
+            "movl %c[state_ebp](%%eax), %%ebp\n"
+            "movups %c[state_xmm0](%%eax), %%xmm0\n" // dump struct regs xmm state
+            "movups %c[state_xmm1](%%eax), %%xmm1\n"
+            "movups %c[state_xmm2](%%eax), %%xmm2\n"
+            "movups %c[state_xmm3](%%eax), %%xmm3\n"
+            "movups %c[state_xmm4](%%eax), %%xmm4\n"
+            "movups %c[state_xmm5](%%eax), %%xmm5\n"
+            "movups %c[state_xmm6](%%eax), %%xmm6\n"
+            "movups %c[state_xmm7](%%eax), %%xmm7\n"
+            "leal %c[real_esp_off](%%esi), %%esi\n" // where will we save the "real" esp?
+            "movl %%esp, (%%esi)\n" // save our esp since we will switch to mcsema esp later
+            "movl %c[state_esp](%%eax), %%esp\n" // switch to mcsema stack
+            //"addl $4, %%esp\n" // we pushed a fake return addr in the CALL{r,m} definitions 
+                               // in x86Instrs_Branches.cpp, undo it
+            "movl %%ecx, 0(%%esp)\n" // use that slot to store jump destination
+            "movl %c[jmp_count](%%esi), %%ecx\n" // save recursion count into ecx
+            "shll $3, %%ecx\n" // multiply recursion count by 8 to get offset (mul 8 = shl 3)
+            "leal 0f, %%esi\n" // base return addr
+            "addl %%ecx, %%esi\n" // calculate return addr
+            "pushl %%esi\n" // push return addr
+            "movl %c[state_ecx](%%eax), %%ecx\n" // complete struct regs spill
+            "movl %c[state_esi](%%eax), %%esi\n" // complete struct regs spill
+            "movl %c[state_eax](%%eax), %%eax\n" // complete struct regs spill
+            "jmpl *4(%%esp)\n"
+            COUNT_LEVEL(0) // set of jump locations that increment the call frame counter
+            COUNT_LEVEL(1) // the amount of these hit depends on the recursion depth
+            COUNT_LEVEL(2) // at depth 0, none are hit, at depth 1, there is 1, etc.
+            COUNT_LEVEL(3) // there are 512 incl entries
+            COUNT_LEVEL(4)
+            COUNT_LEVEL(5)
+            COUNT_LEVEL(6)
```

Comment:
can you put a bunch of these on the same line so that I don't need to scroll so much? :-P